LongestPole/Main.lean

```diff
@@ -0,0 +1,185 @@
+  match Json.parse r with
+  | .error e => throw <| IO.userError s!"Could not parse speed center JSON: {e}\n{r}"
+  | .ok j => match fromJson? j with
+    | .ok v => pure v
+    | .error e => match fromJson? j with
+      | .ok (v : SpeedCenterMessage) =>
+        IO.eprintln s!"http://speed.lean-fro.org says: {v.message}"
+        IO.eprintln s!"Try moving to an older commit?"
+        IO.Process.exit 1
+      | .error _ => throw <| IO.userError s!"Could not parse speed center JSON: {e}\n{j}"
+
+def headSha : IO String := return (← runCmd "git" #["rev-parse", "HEAD"]).trim
+
+def speedCenterInstructions (run : String) : IO (NameMap Float) :=
+  return (← speedCenterRunResponse run).instructions
+
+partial def cumulativeInstructions (instructions : NameMap Float) (graph : NameMap (Array Name)) :
+    NameMap Float :=
+  graph.fold (init := ∅) fun m n _ => go n m
+where
+  go (n : Name) (m : NameMap Float) : NameMap Float :=
+    if m.contains n then
+      m
+    else
+      let parents := graph.find! n
+      let m := parents.foldr (init := m) fun n' m => go n' m
+      let t := (parents.map fun n' => (m.find! n')).foldr max 0
+      m.insert n ((instructions.find? n).getD 0 + t)
+
+def slowestParents (cumulative : NameMap Float) (graph : NameMap (Array Name)) :
+    NameMap Name :=
+  graph.fold (init := ∅) fun m n parents =>
+    match parents.toList with
+    | [] => m
+    | h :: t => Id.run do
+      let mut r := h
+      for n' in t do
+        if cumulative.find! n' > cumulative.find! r then
+          r := n'
+      return m.insert n r
+
+def totalInstructions (instructions : NameMap Float) (graph : NameMap (Array Name)) :
+    NameMap Float :=
+  let transitive := graph.transitiveClosure
+  transitive.filterMap fun n s => some <| s.fold (init := (instructions.find? n).getD 0)
+    fun t n' => t + ((instructions.find? n').getD 0)
+
+open IO.FS IO.Process Name in
+/-- Implementation of the longest pole command line program. -/
+def longestPoleCLI (args : Cli.Parsed) : IO UInt32 := do
+  let to := match args.flag? "to" with
+  | some to => to.as! ModuleName
+  | none => `Mathlib -- autodetect the main module from the `lakefile.lean`?
+  let sha := headSha
+  searchPathRef.set compile_time_search_path%
+  let _ ← unsafe withImportModules #[{module := to}] {} (trustLevel := 1024) fun env => do
+    let graph := env.importGraph
+    IO.eprintln s!"Analyzing {to} at {<- sha}"
+    let instructions ← speedCenterInstructions (← sha)
+    let cumulative := cumulativeInstructions instructions graph
+    let total := totalInstructions instructions graph
+    let slowest := slowestParents cumulative graph
+    let mut table := #[]
+    let mut n := some to
+    while n != none do
+      let i := instructions.find! n.get!
+      let c := cumulative.find! n.get!
+      let t := total.find! n.get!
+      let r := (t / c).toString
+      let r := match r.split (· = '.') with
+      | [a, b] => a ++ "." ++ b.take 2
+      | _ => r
```

Comment:
Can you extract this to some `Float.toStringDecimals (f : Float) (digits : Nat)` helper function?